contrib/ansible-cmdb.man.1
.TH MAN 1 "2015-11-11" "%%VERSION%%" "Ansible CMDB"

.SH NAME
ansible\-cmdb - generate host overview from ansible facts

.SH SYNOPSIS
.B ansible\-cmdb
.RB [options] 
.IR fact_dir\ [fact_dir]

.SH DESCRIPTION

ansible\-cmdb parses facts about machines gathered by Ansible and generates host
overviews or other information based on those facts. It includes the ability to
parse variables, groups and other information from Ansible's hosts inventory
and dynamic inventories.

Output is rendered using templates. At the moment there are various templates
available. You can create your own templates to completely customize
ansible\-cmdb's output.

.SH OPTIONS

Ansible\-cmdb supports the following options:

.SS General options

.TP
.BR \-h,\ \-\-help
Show help information.

.TP
.BR \-\-version
Show version.

.TP
.BR \-d,\ \-\-debug
Show debug information.

.SS Input options

.TP
.BR \-f ", " \-\-fact\-cache =\fIDIR\fR
Use Ansible fact-cache dir for facts

.TP
.BR \-i ", " \-\-inventory =\fIPATH\fR
Scan inventory at PATH for groups, variables, etc.

.SS Output options

.TP
.BR \-t ", " \-\-template =\fITEMPLATE\fR
Template to use. Default is 'html_fancy'.

.TP
.BR \-p ", " \-\-params =\fIKEY=VALUE\fR
Params to send to the template.

.TP
.BR \-c ", " \-\-columns =\fICOL[,COL]\fR
Show only given columns

.TP
.BR \-C ", " \-\-cust\-cols =\fPATH\fR
Add user-defined columns.

.TP
.BR \-l ", " \-\-limit =\fLIMIT\fR
Limit the included hosts to pattern LIMIT

.TP
.BR \-\-exclude\-cols =\fEXCLUDE_COLUMNS\fR
Exclude cols from output

.SH EXAMPLES

First, generate Ansible output for your hosts:

.PP
.nf
.RS
mkdir out \\
ansible -m setup --tree out/ all
.RE
.fi
.PP

To generate a host overview with the default template (html_fancy):

.PP
.nf
.RS
ansible-cmdb out/ > overview.html
.RE
.fi
.PP

To make ansible\-cmdb scan your hosts file, which will include group
information and variables in the host overview:

.PP
.nf
.RS
ansible-cmdb -i /path/to/hosts out/ > overview.html
.RE
.fi
.PP

To select a different template, use the \fB\-t\fR switch. The default template is
html_fancy. If you want to use the txt_table template:

.PP
.nf
.RS
ansible-cmdb -t txt_table out/ > overview.html
.RE
.fi
.PP

If you want to refer to a template at some other location on your disk, you
need to leave the .tpl extension of:

.PP
.nf
.RS
ansible-cmdb -t /home/test/some_template out/ > overview.html
.RE
.fi
.PP

To send parameters to a template, use the \fB\-p\fR swtich. For example: to use
local Javascript resources with the html_fancy template (so you can it
offline):

.PP
.nf
.RS
ansible-cmdb -p local_js=1 -i /path/to/hosts out/ > overview.html
.RE
.fi
.PP

The \fB\-\--columns\fR (\fB\-c\fR ) switch lets you specify which columns
should appear in the overview. This must be supported by the template.

.PP
.nf
.RS
ansible-cmdb --columns name,os,ip,mem,cpus out/ > overview.html
.RE
.fi
.PP

You can specify multiple fact directories. You could, for example, include
hosts that aren't scanned by ansible by manually creating JSON files, or add
additional facts or variables to systems that are also present in the normal
facts dir.

.PP
.nf
.RS
ansible-cmdb out/ manual_facts/ > overview.html
.RE
.fi
.PP


.SH INVENTORY SCANNING

Ansible-cmdb can read your inventory file (hosts, by default), inventory
directory or dynamic inventory and extract useful information from it such as:

.TP
.BR Groups
All the groups a host belongs to. 

.TP
.BR Host\ variables
These are optional key/value pairs for each host which can be used in
playbooks. They are scanned by ansible-cmdb and get added to a hosts discovered
facts under the 'hostvars' section. 

Reading the inventory is done using the \fB\-i\fR switch to ansible-cmdb. It
takes a single parameter: your hosts file, directory containing your hosts
files or path to your dynamic inventory script.

See http://docs.ansible.com/intro_inventory.html#host-variables for more
information on host variables.


.SH TEMPLATES

Ansible\-cmdb supports multiple different templates. You can choose which
template to use with the \fB\-t\fR switch

.SS html_fancy

A fancy HTML page that uses jQuery and DataTables to give you a searchable,
sortable table overview of all hosts with detailed information just a click
away.

It takes a parameter \fIlocal_js=1\fR which will load resources from the local
disk instead of over the network. See the \fBEXAMPLES\fR section for an
example. It also supports the \fB\-\-cols\fR parameter to specify which 
columns you'd like to see in the output.

It can be easily extended by copying it and modifying the cols definition at
the top. 

.SS txt_table

A quick text table summary of the available hosts with some minimal information.

It supports the \fI\-\-cols\fR parameter to specify which columns you'd like in
the output.

.SS json

The json template simple dumps a JSON-encoded representation of the gathered
information. This includes all the extra information scanned by ansible-cmdb
such as groups, variables, custom information, etc.

.SH FACT CACHING

Ansible can cache facts from hosts when running playbooks. This is configured
in Ansible like:\

.PP
.nf
.RS
[defaults]
fact_caching=jsonfile
fact_caching_connection = /path/to/facts/dir
.RE
.fi
.PP

You can use these cached facts as facts directories with ansible\-cmdb by
specifying the \fB\-f\fR (\fB\-\-fact\-cache\fR) option:\

.PP
.nf
.RS
$ ansible-cmdb -f /path/to/facts/dir > overview.html
.RE
.fi
.PP

Please note that the \fB\-\-fact\-cache\fR option will apply to all fact
directories you specify. This means you can't mix fact-cache fact directories
and normal setup fact directories. Also, if you wish to manually extend facts
(see the Extending chapter), you must omit the ansible_facts key and put items
in the root of the JSON.


.SH COLUMNS

Some templates, such as \fBtxt_table\fR and \fBhtml_fancy\fR, support columns.
If a template supports columns, you can use the \fB\-\-columns\fR / \fB\-c\fR
command line option to specify which columns to show.

The \fB\-\-columns\fR takes a comma-separated list of columns (no spaces!)
which should be shown. The columns must be specified by their id field. For
information on what id fields are supported by a template, take a look in the
template.  Usually it's the column title, but in lowercase and with spaces
replaced by underscores.

You can add custom columns with the \fB\-C\fR (\fB\-\-cust\-cols\fR) option.
Please see the online user manual for more information.

.SH EXTENDING

You can specify multiple directories that need to be scanned for output. This
lets you add more custom information to existing hosts or even completely new
hosts.

For example, you could create a directory called out_cust and put manually
crafted JSON files in it:

.PP
.nf
.RS
$ mkdir out_cust
$ cat out_cust/test.megacorp.com
{
  "software": [
    "Apache2",
    "MySQL5.5"
  ]
}
.RE
.fi
.PP

Specify both directories when generating the output:

.PP
.nf
.RS
$ ansible-cmdb out/ out_cust/ > overview.html
.RE
.fi
.PP

our custom variables will be put in the root of the host information dictionary:

.PP
.nf
.RS
"test.megacorp.com": {
  "ansible_facts": {
    "ansible_all_ipv4_addresses": ["185.21.189.140"],
  },
  "changed": false,
  "groups": ["cust.flusso"],
  "software": [
    "Apache2",
    "MySQL5.5"
  ],
  "name": "ad6.flusso.nl"
}
.RE
.fi
.PP

If you're using the \fB\-\-fact\-cache\fR option, you must omit the
ansible_facts key and put items in the root of the JSON. This also means that
you can only extend native ansible facts and not information read from the
hosts file by ansible\-cmdb.

.SH SEE ALSO

\fIREADME.md\fR, the main source of documentation

.SH COPYRIGHT

ansible\-cmdb is copyright 2015 by Ferry Boender

Ferry Boneder
ferry.boender@gmail.com



CHANGELOG.txt
v1.19   2017-01-07

- Always show stack trace on error and include class name.
- Exit with proper exit codes.
- Exclude certain file extensions from consideration as inventories.
- Improved error reporting and lookups of templates.
- Improved error reporting when specifying inventories.

v1.18   2016-12-07

- Python packaging / pip support.
- Fixes in the "Make install" from repository procedure.
- If Ansible-cmdb finds a "hosts" file in the current dir, it will use it.

v1.17   2016-10-08

- A bug that caused whitespace to be added each time a character was typed
  into the search box was fixed.
- Fix for HP-UX missing netmask (patch by Randy Rasmussen)
- SQL template improvements for BSD.

v1.16   2016-09-02

- Added "all ipv6" column to html_fancy template.
- Fixed broken name urls on html_fancy template (patch by Jordan Anderson)

v1.15   2016-07-28

- Improvements to the resilience against wrong, unsupported and missing data.
- SQL template. Generates SQL for use with SQLite and MySQL.
- Minor bugfixes.

v1.14   2016-04-23

- Look for ansible.cfg and use hostfile setting.
- html_fancy: Properly sort vcpu and ram columns.
- html_fancy: Remember which columns the user has toggled.
- html_fancy: display groups and hostvars even if no host information was
  collected.
- html_fancy: support for facter and custom facts.
- html_fancy: Hide sections if there are no items for it.
- html_fancy: Improvements in the rendering of custom variables.
- Apply Dynamic Inventory vars to all hostnames in the group.
- Many minor bugfixes.

v1.13   2016-03-12

- Added csv template: Output a CSV file with host information.
- Added markdown template: Output a Markdown-formatted file with host
  information.
- html_fancy template: Minor cosmetic changes.

v1.12   2016-02-15

- Bugfix: Don't parse hidden files in fact output dirs.
- Bugfix: Various fixes in html_fancy template that prevent output from being
  generated if certain facts are missing.
- Bugfix: Improved error handling to prevent ansible-cmdb from stopping on a
  single invalid host.
- Improved error reporting
- Ansible local facts (from /etc/ansible/facts.d on hosts) is now shown in the
  Host details overview.
- Support undocumented extended YAML vars in hosts files.
- txt_table: Show MemFree and MemUsed columns. (by Kurt Davis)
- txt_table: Properly align table columns if the values are less wide than the
  title.

v1.11   2016-01-12

- Source package improvements in man page handling (Alex Barton)
- html_fancy template now supports OpenBSD facts.
- html_fancy template now supports Windows (2008) facts.
- html_fancy template now shows a link icon instead of a star for the search box URL.
- Improved error reporting and debugging mode (-d).

v1.10   2015-12-21

- html_fancy: Headers are now links
- html_fancy: Search box can now be linked to using the ?search= param, which
  will automatically filter the results when the page is loaded. 
- html_fancy: A star with a link now appears next to the search box for easy
  copy-pasting of the search box URL.
- UTF-8 / unicode bugfixes.

v1.9    2015-12-05 (Happy Sinterklaas)

- Always output UTF-8.
- Added the 'json' template, which dumps the entire inventory including
  variables and groups in JSON format.
- Included a manual page in the packages.
- The -t/--template switch can now point to an actual filename, and no longer
  requires the .tpl extension to be stripped.
- html_fancy: Fixed a bug where the search field was not properly focussed.
- html_fancy: Use vcpus for cores + hyperthreading count (Rowin Andruscavage)
- html_fancy: New cpu, memory, swap columns and improved sorting (Rowin Andruscavage)
- html_fancy: Show a hosts groups in the detailed view
- html_fancy: Move and disable/enable some default columns.
- html_fancy: Network overview table for host details.

v1.8    2015-11-10

- Fixed a bug in Dynamic Inventory parsing
- Spelling mistakes in README fixed (by Sebastian Gumprich)

v1.7    2015-11-08

- Added a -c / --columns option that allows users to specify the columns to
  show. Must be supported by template.
- Html_fancy template now has a kernel column. (by Rowin Andruscavage)
- Html_fancy template now has a timestamp column which displays when facts
  where gathered. (by Rowin Andruscavage)
- Bugfix: Search box in html_fancy template was not properly focused
  automatically.
- Version is now properly displayed with --version
- Development related: Refactoring of ansible-cmdb into a module and a thin
  commandline wrapper.

v1.6    2015-10-03

- The -i switch now supports reading dynamic inventory scripts.
- host_vars directory now supported (by Malcolm Mallardi)
- Support for multiple inventory sources as per
  https://docs.ansible.com/ansible/intro_dynamic_inventory.html#using-multiple-inventory-sources.
- Improved error handling prevents ansible-cmdb from stopping if it encounters
  non-critical errors (malformed host definitions, etc).
- Improved error reporting.
- html_fancy template column headers are now visually identifiable as being
  sortable.

v1.5    2015-09-20

- html_fancy overview page can now be viewed locally (through file:///) without
  needing to specify the -p local_js=1 flag to it. It still fetches resources
  from the internet, so you still require an internet connection. If you do not
  want to fetch resources from the internet, use the -p local_js=1 flag.
- A new Material design for the html_fancy template.

v1.4    2015-08-26

- Support for host inventory patterns (e.g. foo[01:04].bar.com)
- Support for 'vars' and 'children' groups.
- Support passing a directory to the `-i` param, in which case all the files in
  that directory are interpreted as one big hosts file.
- Support for the use of local jquery files instead of via a CDN. Allows you to
  view the hosts overview in your browser using file://. See README.md for info
  on how to enable it (hint: `ansible-cmdb -p local_js=1`).

src/ansible-cmdb.py
#!/usr/bin/env python

# ansible_cmd
#
# Generate host overview (configuration management database) from ansible fact
# gathering output.
#
# Usage:
#
#   $ ansible -m setup --tree out all
#   $ ansible-cmdb out > cmdb.html
#

import optparse
import sys
import os
import logging
import ast
from mako import exceptions
import ansiblecmdb
import ansiblecmdb.util as util
import ansiblecmdb.render as render


# Verify Python version
if sys.version_info < (2, 7):
    sys.stderr.write(
        "Ansible-cmdb requires Python v2.7+. You are running {}. Support for "
        "Python v2.6 supported ended October 2013. You should upgrade to a "
        "newer version.\n".format(sys.version))
    sys.exit(1)


def get_logger():
    """
    Instantiate a logger.
    """
    root = logging.getLogger()
    root.setLevel(logging.WARNING)
    ch = logging.StreamHandler(sys.stderr)
    ch.setLevel(logging.DEBUG)
    formatter = logging.Formatter('%(message)s')
    ch.setFormatter(formatter)
    root.addHandler(ch)
    return root


def get_data_dir():
    """
    Find out our installation prefix and data directory. These can be in
    different places depending on how ansible-cmdb was installed.
    """
    data_dir_paths = [
        os.path.join(os.path.dirname(ansiblecmdb.__file__), 'data'),
        os.path.join(os.path.dirname(sys.argv[0]), '..', 'lib', 'ansiblecmdb', 'data'),
        '/usr/local/lib/ansiblecmdb/data',
        '/usr/lib/ansiblecmdb/data',
    ]

    data_dir = util.find_path(data_dir_paths, 'tpl/html_fancy.tpl')
    if not data_dir:
        sys.stdout.write("Couldn't find the data dir for the templates. I tried: {0}\n".format(", ".join(data_dir_paths)))
        sys.exit(1)

    return data_dir


def get_hosts_files(option):
    """
    Find out the location of the `hosts` file. This looks in multiple places
    such as the `-i` option, current dir and ansible configuration files. The
    first match is returned as a list.
    """
    if option is not None:
        return option.split(',')

    # Use hosts file from the current dir if it exists
    if os.path.isfile('hosts'):
        return ['hosts']

    # Perhaps it's configured in a configuration file. Try to find a
    # configuration file and see if it contains a `hostsfile` entry.
    config_locations = [
        '.',
        '/etc/ansible/'
    ]
    config_dir = util.find_path(config_locations, 'ansible.cfg')
    log.debug('config_dir = {0}'.format(config_dir))
    if config_dir:
        with open(os.path.join(config_dir, 'ansible.cfg'), 'r') as cf:
            for line in cf:
                if line.startswith('hostfile'):
                    return [line.split('=', 1)[1].strip()]


def get_cust_cols(path):
    """
    Load custom column definitions.
    """
    required_keys = ["title", "id", "sType", "visible"]

    with open(path, 'r') as f:
        try:
            cust_cols = ast.literal_eval(f.read())
        except Exception as err:
            sys.stderr.write("Invalid custom columns file: {}\n".format(path))
            sys.stderr.write("{}\n".format(err))
            sys.exit(1)

    # Validate
    for col in cust_cols:
        for required_key in required_keys:
            if required_key not in col:
                sys.stderr.write("Missing required key '{}' in custom "
                                 "column {}\n".format(required_key, col))
                sys.exit(1)
            if "jsonxs" not in col and "tpl" not in col:
                sys.stderr.write("You need to specify 'jsonxs' or 'tpl' "
                                 "for custom column {}\n".format(col))
                sys.exit(1)



    return cust_cols

def parse_user_params(user_params):
    """
    Parse the user params (-p/--params) and them as a dict.
    """
    if user_params:
        params = {}
        try:
            for param in options.params.split(','):
                param_key, param_value = param.split('=', 1)
                params[param_key] = param_value
        except ValueError as e:
            sys.stdout.write("Invalid params specified. Should be in format: <key=value>[,<key=value>..]\n")
            sys.exit(1)
        return params
    else:
        return {}


if __name__ == "__main__":
    log = get_logger()
    data_dir = get_data_dir()
    tpl_dir = os.path.join(data_dir, 'tpl')
    static_dir = os.path.join(data_dir, 'static')
    version = open(os.path.join(data_dir, 'VERSION')).read().strip()

    parser = optparse.OptionParser(version="%prog v{0}".format(version))
    parser.set_usage(os.path.basename(sys.argv[0]) + " [option] <dir> > output.html")
    parser.add_option("-t", "--template", dest="template", action="store", default='html_fancy', help="Template to use. Default is 'html_fancy'")
    parser.add_option("-i", "--inventory", dest="inventory", action="store", default=None, help="Inventory to read extra info from")
    parser.add_option("-f", "--fact-cache", dest="fact_cache", action="store_true", default=False, help="<dir> contains fact-cache files")
    parser.add_option("-p", "--params", dest="params", action="store", default=None, help="Params to send to template")
    parser.add_option("-d", "--debug", dest="debug", action="store_true", default=False, help="Show debug output")
    parser.add_option("-q", "--quiet", dest="quiet", action="store_true", default=False, help="Don't report warnings")
    parser.add_option("-c", "--columns", dest="columns", action="store", default=None, help="Show only given columns")
    parser.add_option("-C", "--cust-cols", dest="cust_cols", action="store", default=None, help="Path to a custom columns definition file")
    parser.add_option("-l", "--limit", dest="limit", action="store", default=None, help="Limit hosts to pattern")
    parser.add_option("--exclude-cols", dest="exclude_columns", action="store", default=None, help="Exclude cols from output")
    (options, args) = parser.parse_args()

    if len(args) < 1:
        parser.print_usage()
        sys.stderr.write("The <dir> argument is mandatory\n")
        sys.exit(1)

    if options.quiet:
        log.setLevel(logging.ERROR)
    elif options.debug:
        log.setLevel(logging.DEBUG)

    hosts_files = get_hosts_files(options.inventory)

    cust_cols = []
    if options.cust_cols is not None:
        cust_cols = get_cust_cols(options.cust_cols)

    # Handle template params
    params = {
        'lib_dir': data_dir,  # Backwards compatibility for custom templates < ansible-cmdb v1.7
        'data_dir': data_dir,
        'version': version,
        'log': log,
        'columns': None,
        'exclude_columns': None,
        'cust_cols': cust_cols
    }
    params.update(parse_user_params(options.params))
    if options.columns is not None:
        params['columns'] = options.columns.split(',')
    if options.exclude_columns is not None:
        params['exclude_columns'] = options.exclude_columns.split(',')

    # Log some debug information
    log.debug('data_dir = {0}'.format(data_dir))
    log.debug('tpl_dir = {0}'.format(tpl_dir))
    log.debug('static_dir = {0}'.format(static_dir))
    log.debug('inventory files = {0}'.format(hosts_files))
    log.debug('template params = {0}'.format(params))

    ansible = ansiblecmdb.Ansible(args, hosts_files, options.fact_cache,
                                  limit=options.limit, debug=options.debug)

    # Render a template with the gathered host info
    renderer = render.Render(options.template, ['.', tpl_dir])
    if renderer.tpl_file is None:
        sys.stderr.write("Template '{0}' not found at any of \n  {1}\n".format(options.template, "\n  ".join(renderer.tpl_possibilities)))
        sys.exit(1)

    # Make sure we always output in UTF-8, regardless of the user's locale /
    # terminal encoding. This is different in Python 2 and 3.
    try:
        output = renderer.render(ansible.get_hosts(), params)
        if output:
            if sys.version_info[0] == 3:
                sys.stdout.buffer.write(output.lstrip())
            else:
                sys.stdout.write(output.lstrip())
    except Exception as err:
        full_err = exceptions.text_error_template().render().replace("\n", "\n    ")
        debug_cmd = "{0} -d {1}".format(sys.argv[0], ' '.join(sys.argv[1:]))
        debug_txt = ("Whoops, it looks like something went wrong while rendering the template.\n\n"
                     "The reported error was: {0}: {1}\n\nThe full error was:{2}\n"
                     "The output is probably not correct.\n\n".format(err.__class__.__name__, err, full_err))
        if err.__class__.__name__ == "KeyError":
            debug_txt += "!!! This is probably a problem with missing information in your host facts\n\n"
        if not options.debug:
            debug_txt += ("You can re-run ansible-cmdb with the -d switch to turn on debugging\n"
                          "to get an insight in what might be going wrong:\n\n"
                          "  {0}\n\n".format(debug_cmd))
        debug_txt += \
"""\
You can report a bug on the issue tracker:

  https://github.com/fboender/ansible-cmdb/issues

Please include the debugging output (-d switch) in the report!

If you can, also include the hosts file and the facts file for the last host
that rendered properly ('Rendering host...' in the output. If these files must
remain confidential, you can send them to ferry.boender@gmail.com instead.
"""
        sys.stderr.write(debug_txt)
        sys.exit(1)
